scripts/tools/generate_esp32_chip_factory_bin.py

```diff
@@ -0,0 +1,175 @@
+import cryptography.x509
+from types import SimpleNamespace
+
+if os.getenv('IDF_PATH'):
+    sys.path.insert(0, os.path.join(os.getenv('IDF_PATH'),
+                                    'components',
+                                    'nvs_flash',
+                                    'nvs_partition_generator'))
+    import nvs_partition_gen
+else:
+    log.error("Please set the IDF_PATH environment variable.")
+    exit(0)
+
+INVALID_PASSCODES = [00000000, 11111111, 22222222, 33333333, 44444444, 55555555,
+                     66666666, 77777777, 88888888, 99999999, 12345678, 87654321]
+
+TOOLS = {}
+
+
+def check_tools_exists():
+    TOOLS['spake2p'] = shutil.which('spake2p')
+    if TOOLS['spake2p'] is None:
+        logging.error('spake2p not found, please add spake2p path to PATH environment variable')
+        sys.exit(1)
+
+
+def validate_args(args):
+    # Validate the passcode
+    if args.passcode is not None:
+        if ((args.passcode < 0x0000001 and args.passcode > 0x5F5E0FE) or (args.passcode in INVALID_PASSCODES)):
+            logging.error('Invalid passcode:' + str(args.passcode))
+            sys.exit(1)
+
+    # Validate the discriminator
+    if (args.discriminator is not None) and (args.discriminator not in range(0x0000, 0x0FFF)):
+        logging.error('Invalid discriminator:' + str(args.discriminator))
+        sys.exit(1)
+
+    logging.info('Discriminator:{} Passcode:{}'.format(args.discriminator, args.passcode))
+
+
+def gen_spake2p_params(passcode):
+    iter_count_max = 10000
+    salt_len_max = 32
+
+    cmd = [
+        TOOLS['spake2p'], 'gen-verifier',
+        '--iteration-count', str(iter_count_max),
+        '--salt-len', str(salt_len_max),
+        '--pin-code', str(passcode),
+        '--out', '-',
+    ]
+
+    output = subprocess.check_output(cmd)
+    output = output.decode('utf-8').splitlines()
+    return dict(zip(output[0].split(','), output[1].split(',')))
+
+
+def gen_raw_ec_keypair_from_der(key_file, pubkey_raw_file, privkey_raw_file):
+    with open(key_file, 'rb') as f:
+        key_data = f.read()
+
+    key_der = cryptography.hazmat.primitives.serialization.load_der_private_key(key_data, None)
+
+    private_number_val = key_der.private_numbers().private_value
+    with open(privkey_raw_file, 'wb') as f:
+        f.write(private_number_val.to_bytes(32, byteorder='big'))
+
+    public_key_first_byte = 0x04
+    public_number_x = key_der.public_key().public_numbers().x
+    public_number_y = key_der.public_key().public_numbers().y
+    with open(pubkey_raw_file, 'wb') as f:
+        f.write(public_key_first_byte.to_bytes(1, byteorder='big'))
+        f.write(public_number_x.to_bytes(32, byteorder='big'))
+        f.write(public_number_y.to_bytes(32, byteorder='big'))
+
+
+def generate_nvs_bin(args, spake2p_params):
+    dac_raw_privkey = 'dac_raw_privkey.bin'
+    dac_raw_pubkey = 'dac_raw_pubkey.bin'
+    gen_raw_ec_keypair_from_der(args.dac_key, dac_raw_pubkey, dac_raw_privkey)
```

Comment:
It is dangerous to keep all the private keys unprotected. Usually key packages are managed in a way that requires authentication/passkeys to get access to the bits, and are treated very carefully to avoid leaving them around, so that they cannot be leaked. **Leaking of DAC private keys may lead to attestation chain revokation**. Please add comments and warning prints about being careful with those keys. Otherwise, some customers may cut'n'paste and not treat these with caution.

Reply:
Thank you for pointing this out, I have added warnings prints and comments.